x-pack/plugins/security/server/plugin.ts

```diff
@@ -259,6 +268,13 @@ export class SecurityPlugin
     const config = this.getConfig();
     const kibanaIndexName = this.getKibanaIndexName();
 
+    async function getUserId(request: KibanaRequest): Promise<string | undefined> {
+      const [, , security] = await core.getStartServices();
+      return security.authc.getCurrentUser(request)?.profile_uid;
+    }
+
+    dataViews.setGetUserId(getUserId);
+
```

Comment:
__question:__ I assume here we're trying to workaround a cyclic dependency between Security and DataViews plugins? If so, can you try to make `Security` a [runtime dependency](https://github.com/elastic/kibana/pull/167113) of `DataViews` instead?

```json
{
  "type": "plugin",
  "id": "@kbn/data-views-plugin",
   ...
  "plugin": {
    "id": "dataViews",
    ...
    "runtimePluginDependencies" : ["security"]
  }
}
```
```ts
// DataViews.start
plugins.onStart<{ security: SecurityPluginStart }>('security').then(({ security }) => {
  if (security.found) {
    // security.contract.authc.getCurrentUser;
  }
});
```

Reply:
This is an impediment - `For instance, cross-plugin type imports are not technically possible at the moment, given that plugins are referencing each others via TS refs, and refs forbid cyclic dependencies. Which means that to leverage this to address cyclic dependency issues, the public types of at least one of the two plugins will have to be extracted to a shared place (likely a package).`

I just assume keep the existing plugin contract an address this later.

Reply:
@azasypkin Then I need some guidance on where the Security plugin start contract should live - it needs to be in a package - and how much needs to be moved at this point. I'm only interested in one function but you might want to see more moved.

Reply:
> @azasypkin Then I need some guidance on where the Security plugin start contract should live 

Sure, I can help. But my idea was just to replace `DataViews.setGetUserId` with something like that:

```ts
// DataViews.start
plugins.onStart<{ security: SecurityPluginStart }>('security').then(({ security }) => {
  if (security.found) {
    this.getUserId = (request) => security.authc.getCurrentUser(request)?.profile_uid;
  } else {
    // Throw or log a warning if `getUserId` is optional...
  }
});
```

With this approach, you wouldn't have to tweak the Security plugin at all, and all the changes stay within the DataViews plugin, unless I'm missing something?

I'm also wondering about DataViews' behavior when `profile_uid` isn't present. The profile ID is accessible only for users logged in with a valid session in Kibana. However, when accessing Kibana APIs through cURL or another HTTP client, and providing credentials via HTTP headers, the profile ID might not be available. How should DataViews handle this scenario?

Reply:
>With this approach, you wouldn't have to tweak the Security plugin at all, and all the changes stay within the DataViews plugin, unless I'm missing something?

It would still create a cyclical dependency due to the `SecurityPluginStart` import.

Reply:
>However, when accessing Kibana APIs through cURL or another HTTP client, and providing credentials via HTTP headers, the profile ID might not be available. How should DataViews handle this scenario?

The DataViews plugin needs a unique user identifier. It sounds like its missing in some cases. I guess its unclear to me why the profile ID isn't available with these other auth methods.

Reply:
> It would still create a cyclical dependency due to the SecurityPluginStart import.

That's what I'm not sure about, as Kibana didn't complain when I tried it locally. @pgayvallet, is importing **types** of the runtime dependency (`import type { SecurityPluginStart } from '@kbn/security-plugin/server';`) treated as a circular dependency? Context: Security depends on DataViews, and we're exploring ways to use Security from DataViews via runtime dependency (creating a dedicated package to overcome this feels a bit more complicated comparing to a runtime dependency).

Reply:
> I guess its unclear to me why the profile ID isn't available with these other auth methods.

It's a current limitation, not an easy one to overcome. In a nutshell, the profile ID gets fetched and stored in a user session during what we call the 'profile activation' stage in Elasticsearch. This occurs only during login when the user provides their credentials.

Now, when you access Kibana via an HTTP client like `curl`, there's no login event or session since credentials are expected with each request. Activating the profile for every request would be too costly, that's why we don't have profile ID in this case.

However, after going through the PR description, it's not immediately apparent whether there's a need to support caching for programmatic access too, or if focusing on supporting user interactive access would suffice?

Reply:
@azasypkin This is the commit that gave me the circ dependency - https://github.com/elastic/kibana/pull/168910/commits/42bd2f41f21bf49031bef64b4c44531486c60660 - I guess I'm not sure why some of the type checks can give different results locally and in CI.

I'm having trouble thinking of a case where a kibana HTTP request would need a user session but I might be lacking in imagination. Our need for this is entirely based on clearing the browser cache when there's a change in user login.

Reply:
I created https://github.com/elastic/kibana/pull/171533 which reapplies that commit

Reply:
I apologize for the late reply. The dynamic dependency feature we recently added only address the runtime part of cross dependencies between plugins.

For types, it's another story, as we're using TS refs to build our plugins and as TS refs do not allow cyclic/cross dependencies.

In that regard, the only solution would be to extract either one or both plugin's types to packages, to break the type dependency. This is for example what Core has done with their domain package (we have public type packages and then internal implementation packages).

It's indeed some work though.

Reply:
@pgayvallet I'd prefer to avoid that for now and continue with the PR as is.

Reply:
Thanks @pgayvallet ! Let me give a "separate package with Security types" a go, and see how it goes. I'll report back here.

Reply:
Created a draft PR with Security Plugin types: https://github.com/elastic/kibana/pull/171886

Reply:
Okay, https://github.com/elastic/kibana/pull/171886 is merged now, can you please try switching to a runtime dependency again. This time you'll need to import Security plugin types from `kbn-security-plugin-types-{common|public|server}` instead.